File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    visited = [False] * len(dungeons)
    rarr = []
    def dfs(k, d, c):
        if d >= len(dungeons):
            rarr.append(c)
            return
        
        for i in range(len(dungeons)):
            if not visited[i]:
                visited[i] = True
                need = dungeons[i][0]
                use = dungeons[i][1]
                if k >= need:
                    dfs(k - use, d + 1, c + 1)
                else :
                    dfs(k, d + 1, c)
                visited[i] = False
    dfs(k,0,0)
    return max(rarr)

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    boolean[] visited;
    int max = 0;
    
    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        dfs(0, k, dungeons);
        return max;
    }
    
    public void dfs(int depth, int k, int[][] dungeons){
        max = Math.max(max, depth);
        for(int i=0; i<dungeons.length;i++){
            int r = dungeons[i][0];
            int c = dungeons[i][1];
            if(!visited[i] && k >= r){
                visited[i] = true;
                dfs(depth + 1, k - c, dungeons);
                visited[i] = false;
            }
        }
        
    }
    
    
}
